README
Finance-QuoteHist 
-----------------

The Finance-QuoteHist bundle is several modules designed to fetch
historical stock quotes from the web: 
  
Finance::QuoteHist
------------------
Top level aggregator that will select a default lineup of
site instances from which to retrieve quotes. Other
than the default lineup, this module behaves identically
to whichever site-specific module is first in the lineup.
See below for site-specific modules. 

Finance::QuoteHist::Generic
---------------------------
The real workhorse of the bundle, a sub class of
LWP::UserAgent. Site-specific modules are sub classes
of this module.

Current site-specific modules include:

Finance::QuoteHist::Yahoo
Finance::QuoteHist::Yahoo::Australia
Finance::QuoteHist::QuoteMedia
Finance::QuoteHist::Google
Finance::QuoteHist::BusinessWeek
Finance::QuoteHist::MSN


INSTALLATION

You install HTML-Element-Extended, as you would install any
perl library, by running these commands:

   perl Makefile.PL
   make
   make test
   make install


DOCUMENTATION

POD style documentation is included with each module. This is normally
converted to a manual page and installed as part of the "make install"
process. You should also be able to use the 'perldoc' utility to
extract and read documentation from the module file directly.  See
Changes for recent changes.


SUPPORT

There is a mailing list available. To view past messages or subscribe,
go to:

   http://lists.sourceforge.net/mailman/listinfo/quotehist-general

Questions and comments can also be directed to Matt Sisk <sisk@mojotoad.com>

AVAILABILITY

The library is available from CPAN:

   http://www.cpan.org/authors/id/M/MS/MSISK/

The latest version is also available at:

   http://www.mojotoad.com/sisk/projects/Finance-QuoteHist/


DISCLAIMER 

The data returned from these modules is in no way guaranteed, nor
are the developers responsible in any way for how this data (or lack
thereof) is used. The interface is based on URLs and page layouts
that might change at any time. Even though these modules are designed
to be adaptive under these circumstances, they will at some point
probably be unable to retrieve data unless fixed or provided with new
parameters. Furthermore, the data from these web sites is usually not
even guaranteed by the web sites themselves, and oftentimes is
acquired elsewhere. See the documentation for each site-specific
module for more information regarding the disclaimer for that site. 


ACKNOWLEDGEMENTS

Thanks to Jim Miner for his generous efforts testing this module. Thanks
to Josh Woodward for some pointers on the Yahoo! interface. Thanks to
Jacob Anawalt for pointers on the Yahoo! interface and suggestions on
open-ended date queries. Thanks to Paul McDermott and Jay Strauss for
patches. Thanks to Manoj Bhatti, Bernd Hacker, Christian von
Engelbrechten, Gary Nielson, Mike Brown, Bill Stephenson, and Robert S.
Weigel for their bug-spotting.


COPYRIGHT

Copyright (c) 1999-2011 Matthew P. Sisk.  All rights reserved.
All wrongs revenged. This program is free software; you can
redistribute it and/or modify it under the same terms as
Perl itself.


Changes
Revision history for Perl extension Finance::QuoteHist.

1.27  Tue Jun 27 14:19:13 EDT 2017
    - Fixed Yahoo module to handle the site's obfuscation
    - updated test data
    - removed defunct StockNod

1.24  Mon Aug  3 15:55:29 EDT 2015
    - removed some deprecated Date::Manip config statements
    - removed outmoded fetchers

1.23  Thu May 21 11:48:11 EDT 2015
    - fixed divident/split entanglement for yahoo
    - patched quotes test to properly skip tests
      without network connectivity (RT #104358)

1.22  Mon Apr 20 14:15:32 EDT 2015
    - fixed dividend/split entanglement on yahoo
      rt.cpan.org #103759

1.21  Fri Jan  9 11:25:10 EST 2015
    - fixed splits on yahoo
    - removed BusinessWeek

1.20  Tue Oct  8 14:54:48 EDT 2013
    - added DailyFinance
    - removed QuoteMedia, MSN, Investopedia

1.19  Mon Aug 29 11:08:08 EDT 2011
    - fixed test data window (1.17 was a little *too*
      recent for reliable monthly tests)
    - 1.18 skipped, CPAN version appeasement

1.17  Fri Aug 26 19:31:05 EDT 2011
    - Updated tests with more recent data

1.16  Thu Sep 30 01:00:27 EDT 2010
    - Fixed tests so that they no longer rely upon historical adjusted
      quotes since these appear to change every now and then whenever
      Yahoo decides to tweak their algorithm.
    - typo fix

1.14  Mon Jun  7 18:36:30 EDT 2010
    - Fixed several syntax issues that were tripping up -w
    - Fixed the utf8 Byte Order Mark issues with google csv
    - Added modules for Investopedia.com and StockNod.com
    - Expanded site-specific tests (for development mostly)
    - Test data updated, docs refreshed and tweaked

1.12  Mon Aug  3 18:40:15 EDT 2009
    - Fixed various site-specific modules
    - F::Q::Yahoo_AU is deprecated in favor of
      F::Q::Yahoo::Australia (though the old package
      is still around for backwards compat)
    - moved CSV class discovery to package level (thanks
      to Edmond Shum)
    - fixed the 'unsigned integer' problem with volumes
      (thanks to 'o kite')
    - fixed some undef warnings and typos (thanks to
      Jim Miner)

1.11  Mon Jun 25 19:21:21 EDT 2007
    - Added Google and Yahoo_AU
    - Rewrote BusinessWeek, decodes from javascript channels now,
      since they changed how their site operates. No more intraday
      quotes though.

1.10  Tue Mar 13 13:57:07 EDT 2007

    - Tweaked Yahoo test data yet again (they seem to be rounding off
      their volumes now)
    - Fixed undefined error for symbols with no closing (thanks to
      Terry Gliedt)
    - Added BusinessWeek to lineup, including intraday functionality.
    - Flattened champion delegation.
    - Added labels() documentation
    - Implemented user-defined row filtering callback (thanks to Manoj
      Bhatti for the suggestion)
    - Added 'shares' as possible label for 'vol' (thanks again to Manoj)

1.09  Thu Jun  8 16:19:57 EDT 2006
    - Fixed ^DJI test data (again) since yahoo keeps changing how they
      represent volume (thanks to Ivo Welch)

1.08  Sat May  6 15:16:11 EDT 2006
    - Fixed ^DJI test data for quote tests (Yahoo switched to the volume
      for each anchor date rather than aggregate volumes)
    - splits() now returns ref properly in scalar context for
      F::QH::Yahoo. Thanks to Raymond de Leon for spotting.

1.07  Thu Mar 16 00:44:24 EST 2006
    - Internal restructure to date iterators where appropriate.
    - Doc fix for MSN.
    - Various minor edge cases fixed.

1.06  Fri Mar 10 15:59:05 EST 2006
    - Added MSN quote source.
    - Fixed some edge cases involving null-queries on splits.
    - Various minor fixes, including proper business-day calculations
      for end of month dates on montly resolutions, plus proper newline
      stripping (chomp didn't work on MSN csv).

1.05  Tue Feb 28 16:47:57 EST 2006
    - Splits now have no dependencies on dividend queries
    - Daily, Weekly, Monthly granularities available
    - Restored quote_source(), dividend_source(), and
      split_source() methods
    - More tests added

1.04  Fri Jan  6 12:18:59 EST 2006
    - Compensated for yahoo symbol switch from ^DJA to ^DJI, which was
      causing some tests to fail.
    - Updated quotemedia url.

1.03  Wed Nov  2 16:52:30 EST 2005
    - Minor bug fixes
    - Forced date calcs to GMT to avoid some cygwin/win errors
    - Added auto_proxy, enabled during tests so that if $ENV{HTTP_PROXY}
      is present $ua->env_proxy will be automatically invoked
    - Tests will skip if network connectivity cannot be established

1.02  Wed Aug  3 14:46:43 EDT 2005
    - Fixed Date::Manip requirement that was inadvertently dropped.

1.01  Thu Jul 21 13:29:54 EDT 2005
    - Added conditional support for Text::CSV_PP in cases where
      Text::CSV_XS is not installed
    - Split out tests

1.00  Thu Jun 23 16:25:53 EDT 2005
    - Reworked Yahoo! queries for new interface
    - Added quotes from QuoteMedia (they supply SiliconInvestor, so that
      was dropped)
    - Date constraints relaxed. No end date defaults to 'today', no
      start date defaults to inception. Having neither date grabs
      entire history.
    ! Internal API changed from symbolic dereferencing to supplying
      closures and iterators for particular targets and parse modes.

0.33
    - Broke up Yahoo! csv queries into blocks of 1000, as apparently
      Yahoo! sometimes puts a cap on these sort of queries (thanks to
      Jay Strauss)

0.32  Thu Aug 14 13:37:36 EDT 2003
    - Fixed csv parsing for Yahoo dividend data after they fixed
      their format
    - Updated test data

0.31  Thu Jan 16 19:20:31 GMT 2003
    - Rolled back some 5.6 specific syntax

0.30  Fri Nov  8 16:47:07 EST 2002
    - adjusted volume fix (column labeling snafu)
    - extraneous status messages fixed for split extrations
    - added filtered split and filtered dividend tests

0.29  Thu Oct 17 04:36:05 EDT 2002
    - Patched Yahoo URL interface since they switched to January = 00
      rather than 01.


lib/Finance/QuoteHist/BusinessWeek.pm
package Finance::QuoteHist::BusinessWeek;

use strict;
use vars qw(@ISA $VERSION);
use Carp;

$VERSION = '1.02';

use constant DEBUG => 0;

use Finance::QuoteHist::Generic;
@ISA = qw(Finance::QuoteHist::Generic);

use MIME::Base64;
use Date::Manip;
Date::Manip::Date_Init("TZ=GMT");

# Example URL:
#
# /businessweek/research/common/charts/update_historical_chart.asp?duration=3653&frequency=1day&freq=undefined&scaling=linear&display=mountain&exportData=1&dMin=37431&dMax=39257&uppers=[]&lowers=[]&lastColorUsed=2&action=&action_value= H

my %Label_Map = (
  d => 'date',
  h => 'high',
  c => 'close',
  v => 'volume',
  l => 'low',
  o => 'open',
);
my @Sorted = qw(date open high low close volume);

sub new {
  my $that = shift;
  my $class = ref($that) || $that;
  my %parms = @_;
  my $self = __PACKAGE__->SUPER::new(%parms);
  bless $self, $class;
  $self->parse_mode('businessweek_javascript');
  $self->labels(
    target_mode => 'quote',
    labels => [ qw(d h c v l o) ],
  );
  $self;
}

sub businessweek_javascript_parser {
  my $self = shift;
  sub {
    my $data = shift;
    return [] unless defined $data;

    my %columns;
    my @pats = sort keys %Label_Map;
    foreach my $line (grep(/arrayDecodeUnpack/i, split(/\s*;\s*/, $data))) {
      next unless $line =~ /(\w+)=arrayDecodeUnpack\s*\(\"([^\"]+)\"\s*\)/i;
      my($varname, $value) = ($1, $2);
      my $label;
      foreach my $pi (0 .. $#pats) {
        my $pat = $pats[$pi];
        if ($varname =~ /^($pat)/i) {
          $label = $Label_Map{$pat};
          splice(@pats, $pi, 1);
          last;
        }
      }
      if (! $label) {
        print STDERR "skipping unknown label $varname\n" if DEBUG;
        next;
      }
      if ($columns{$label}) {
        warn "oops, seen $label, skipping\n" if DEBUG;
        next;
      }
      $columns{$label} = $value;
    }
    my $count = 0;
    foreach my $label (keys %columns) {
      my $decoded = _decode_unpack($columns{$label});
      print STDERR "$label : ", scalar @$decoded, " items\n" if DEBUG;
      $count = @$decoded if @$decoded > $count;
      $columns{$label} = $decoded;
    }
    my @rows;
    foreach my $i (0 .. $count-1) {
      my @row;
      foreach my $label (@Sorted) {
        my $value = $columns{$label}[$i];
        $value = $self->excel_daynum_to_date($value) if $label eq 'date';
        push(@row, $value);
      }
      push(@rows, \@row);
    }
    \@rows;
  };
}

sub granularities { qw( daily weekly monthly ) }

sub url_maker {
  my($self, %parms) = @_;
  my $target_mode = $parms{target_mode} || $self->target_mode;
  my $parse_mode  = $parms{parse_mode}  || $self->parse_mode;

  # *always* block uknown target mode and parse mode combinations in
  # order for cascade to work!
  return undef unless
    ($target_mode eq 'quote') && $parse_mode eq 'businessweek_javascript';

  my $granularity = lc($parms{granularity} || $self->granularity);
  my $grain;
  my ($g) = $granularity =~ /^\s*(\w)/;
  if    ($g eq 'd') { $grain = '1day'   } # daily
  elsif ($g eq 'w') { $grain = '1week'  } # weekly
  elsif ($g eq 'm') { $grain = '1month' } # monthly
  else  { croak "Unknown granularity '$granularity'\n" }

  my($ticker, $start_date, $end_date) =
    @parms{qw(symbol start_date end_date)};
  $start_date ||= $self->start_date;
  $end_date   ||= $self->end_date;

  # businessweek has a one-off error on the start date when using weekly
  # or monthly mode
  if ($self->granularity =~ /^(m|w)/i) {
    if ($1 eq 'w') {
      $start_date = DateCalc($start_date, '- 1 week');
    }
    else {
      $start_date = DateCalc($start_date, '- 1 month');
    }
  }

  my $host = 'investing.businessweek.com';
  my $path = 'businessweek/research/common/charts/update_historical_chart.asp';

  # this also worked
  # my $path = '/research/stocks/snapshot/historical.asp';

  my %query = (
    symbol        => $ticker,
    freq          => $grain,
    dMin          => $self->date_to_excel_daynum($start_date),
    dMax          => $self->date_to_excel_daynum($end_date),
    # dunno what the rest of these do
    duration      => 1096,
    frequency     => '1day', # relic?
    scaling       => 'linear',
    display       => 'mountain',
    exportData    => '1',
    uppers        => '[]',
    lowers        => '[]',
    lastColorUsed => 2,
    action        => '',
    action_value  => '',
  );


  my $url = "http://$host/$path?" .
            join('&', map { "$_=$query{$_}" } sort keys %query);

  print STDERR "URL: $url\n" if DEBUG;

  my @urls = ($url);
  sub { pop @urls };
}

### subroutines to deal with businessweek's wonky data encoding

sub _decode_unpack { _array_unpack(MIME::Base64::decode_base64(shift)) }

sub _array_unpack {
  # inspired by this (who comes up with this stuff?):
  # http://investing.businessweek.com/includes/asplib/ArrayUnpack.js
  my $str = shift;
  my @e;
  my @code = map { ord($_) } split(//, $str);
  my $m = $code[1] | ($code[2] << 8);
  if ($m > 0) {
    my $b = $code[3];
    my $f = 10 ** $code[4];
    my $k = $code[5] | ($code[6] << 8) | ($code[7] << 16) | ($code[8] << 24);
    my $l = (2 ** $b) - 1;
    push(@e, $k/$f);
    my($g, $h) = (0, 0);
    for (my $i = 9; $i < @code; $i++) {
      $g |= $code[$i] << $h;
      $h += 8;
      while ($h >= $b && @e <= $m) {
        my $j = ($g & $l) >> 1;
        $k += ($g & 1) ? -1 * $j : $j;
        push(@e, $k/$f);
        $g >>= $b;
        $h -= $b;
      }
      $g |= $code[$i] >> (8 - $h);
    }
  }
  wantarray ? @e : \@e;
}

### excel daynum conversions

# using locale TZ
#sub excel_daynum_to_date {
#  my $self = shift;
#  my $excel_daynum = shift;
#  my $skew = _excel_daynum_sub_skew($excel_daynum);
#  my $date = _dm_from_epoch(_days_to_secs($skew));
#  my $tzod = _tz_offset_in_days($date);
#  _dm_from_epoch(_days_to_secs($skew - $tzod));
#}

# using GMT
sub excel_daynum_to_date {
  my $self = shift;
  my $excel_daynum = shift;
  my $skew = _excel_daynum_sub_skew($excel_daynum);
  my $date = _dm_from_epoch(_days_to_secs($skew));
  _dm_from_epoch(_days_to_secs($skew));
}

# using locale TZ
#sub date_to_excel_daynum {
#  my $self = shift;
#  my $date = shift;
#  _secs_to_days(Date::Manip::UnixDate($date, "%s")) +
#    (_excel_daynum_add_skew(_tz_offset_in_days($date)));
#}

# using GMT
sub date_to_excel_daynum {
  my $self = shift;
  my $date = shift;
  _excel_daynum_add_skew(_secs_to_days(Date::Manip::UnixDate($date, "%s")));
}

# subroutines

sub _dm_from_epoch {
  my $epoch = shift;
  Date::Manip::ParseDateString("epoch $epoch");
}

# using locale TZ
#sub _tz_offset_in_minutes {
#  my $date = shift;
#  my $tz_offset = Date::Manip::UnixDate($date, "%z");
#  my($sign, $hours, $minutes) = $tz_offset =~ /^(\D?)(\d\d)(\d\d)/;
#  $tz_offset = ($hours*60 + $minutes);
#  $tz_offset *= -1 if $sign eq '-';
#  $tz_offset;
#}

# using locale TZ
#sub _tz_offset_in_days { _tz_offset_in_minutes(@_) / (60 * 24) }

sub _excel_daynum_sub_skew { shift() - 25569 }
sub _excel_daynum_add_skew { shift() + 25569 }

sub _days_to_secs { shift() * 86400 }
sub _secs_to_days { shift() / 86400 }

1;

__END__

=head1 NAME

Finance::QuoteHist::BusinessWeek - Site-specific class for retrieving historical stock quotes.

=head1 SYNOPSIS

  use Finance::QuoteHist::BusinessWeek;
  $q = Finance::QuoteHist::BusinessWeek->new
     (
      symbols    => [qw(IBM UPS AMZN)],
      start_date => '01/01/2005',
      end_date   => 'today',
     );

  foreach $row ($q->quotes()) {
    ($symbol, $date, $open, $high, $low, $close, $volume) = @$row;
    ...
  }

=head1 DESCRIPTION

Finance::QuoteHist::BusinessWeek is a subclass of
Finance::QuoteHist::Generic, specifically tailored to read historical
quotes from the Business Week Online web site
(I<http://investments.businessweek.com/>).

BusinessWeek offers granularities of daily, weekly, or monthly.

BusinessWeek does not currently provide information on dividends or splits.

Please see L<Finance::QuoteHist::Generic(3)> for more details on usage
and available methods. If you just want to get historical quotes and are
not interested in the details of how it is done, check out
L<Finance::QuoteHist(3)>.

=head1 METHODS

The basic user interface consists of a single method, as shown in the
example above. That method is:

=over

=item quotes()

Returns a list of rows (or a reference to an array containing those
rows, if in scalar context). Each row contains the B<Symbol>, B<Date>,
B<Open>, B<High>, B<Low>, B<Close>, and B<Volume> for that date. Quote
values are pre-adjusted for this site.

=back

=head1 REQUIRES

Finance::QuoteHist::Generic(3), MIME::Base64(3), Date::Manip(3)

=head1 DISCLAIMER

The data returned from these modules is in no way guaranteed, nor are
the developers responsible in any way for how this data (or lack
thereof) is used. The interface is based on URLs and page layouts that
might change at any time. Even though these modules are designed to be
adaptive under these circumstances, they will at some point probably be
unable to retrieve data unless fixed or provided with new parameters.
Furthermore, the data from these web sites is usually not even
guaranteed by the web sites themselves, and oftentimes is acquired
elsewhere.

Details for BusinessWeek's terms of use can be found here:
I<http://www.businessweek.com/copyrt.htm>

If you still have concerns, then use another site-specific historical
quote instance, or none at all.

Above all, play nice.

=head1 AUTHOR

Matthew P. Sisk, E<lt>F<sisk@mojotoad.com>E<gt>

=head1 COPYRIGHT

Copyright (c) 2006-2010 Matthew P. Sisk. All rights reserved. All wrongs
revenged. This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=head1 SEE ALSO

Finance::QuoteHist::Generic(3), Finance::QuoteHist(3), perl(1).

=cut
